### DISCUSSION TITLE
Why are api-route First Load JS going from 0b to 109kb after upgrade to NextJs 15?

### DISCUSSION THREAD

**Discussion by J4v4Scr1pt** [vercel/next.js#73986] [Help] [answered]
### Summary

After Upgrading to NextJs 15 and React 19 my build went from this:
![image](https://github.com/user-attachments/assets/6e8d6915-6bb9-4e37-985b-9da0cf2a5491)

To this:
![image](https://github.com/user-attachments/assets/0a9ffe49-4837-4cb7-b745-ba2e3a38d53a)

Do not understand the reason, help understanding would be much appreciated :)!

**Comment by icyJoseph** [comment]
Could you help us debug this, please install, https://www.npmjs.com/package/@next/bundle-analyzer, and run it. It'll open three browser windows, one of them has what'll be sent to the client, could you post a screenshot of that?

**Reply by icyJoseph** [reply]
Mmm, I was able to quickly reproduce it:

```console
Route (app)                              Size     First Load JS
┌ ○ /                                    5.47 kB         111 kB
├ ○ /_not-found                          982 B           106 kB
└ ƒ /foo                                 135 B           105 kB
+ First Load JS shared by all            105 kB
  ├ chunks/4bd1b696-92810b4b4ece63ad.js  53 kB
  ├ chunks/517-c94eb82a0c6a5f4b.js       50.4 kB
  └ other shared chunks (total)          1.87 kB
```

**Reply by icyJoseph** [reply]
My first thoughts are that the reporting tool is making a mistake, and that the JS is not part of the bundle, but I'll have to continue tomorrow.

**Reply by J4v4Scr1pt** [reply]
Thanks @icyJoseph !
Sry I missed your first message. Let me know if there is anything I can do to help!

**Reply by J4v4Scr1pt** [reply]
![image](https://github.com/user-attachments/assets/2df9b77a-6066-44c3-a073-8c13089de761)

**Reply by J4v4Scr1pt** [reply]
The three biggest chunks to the left.
![image](https://github.com/user-attachments/assets/1584a1d8-e190-4da1-bf36-27096aa5a4cb)
![image](https://github.com/user-attachments/assets/38ac5983-0bf3-4930-823d-047114c9e917)
![image](https://github.com/user-attachments/assets/13eec998-0a53-41c4-a7fb-a0c4adcf35f0)

**Reply by J4v4Scr1pt** [reply]
Build with Nextjs 14.2.20 and React(+dom) 18.3.1:
![image](https://github.com/user-attachments/assets/b946cc80-9d2f-4bfe-94ec-8cf137e2e8db)

**Reply by J4v4Scr1pt** [reply]
Three biggest chunks to the left with old build:
![image](https://github.com/user-attachments/assets/35866f7d-76c3-4893-b4d6-67949e5aa8d3)
![image](https://github.com/user-attachments/assets/4dea4ace-b3d4-4d1f-820a-4a8bdcc14636)
![image](https://github.com/user-attachments/assets/6d1b3a90-6d3d-4e49-83e7-247c20c54145)

**Reply by J4v4Scr1pt** [reply]
I tried to revert back to NextJs 14.2.20 and the output went back to normal with 0kb on the api routes.

**Reply by J4v4Scr1pt** [reply]
If you need to check anything here is the package.json:

```
{
	"name": "my-app",
	"version": "0.1.0",
	"private": true,
	"type": "module",
	"scripts": {
		"preinstall": "npx only-allow pnpm",
		"predev": "rm -fr .next",
		"dev": "next dev --turbopack",
		"build": "next build",
		"start": "next start",
		"lint": "next lint",
		"ts-lint": "npx tsc",
		"analyze": "cross-env ANALYZE=true next build",
		"dev:debug": "cross-env DEBUG=next:* next dev"
	},
	"dependencies": {
		"@internationalized/date": "3.6.0",
		"@launchdarkly/node-server-sdk": "9.7.2",
		"@nextui-org/react": "2.6.8",
		"@nivo/bar": "0.88.0",
		"@nivo/core": "0.88.0",
		"@nivo/pie": "0.88.0",
		"@tanstack/react-query": "5.62.7",
		"@my/components": "0.9.15",
		"framer-motion": "11.15.0",
		"i18next": "24.1.2",
		"i18next-browser-languagedetector": "8.0.2",
		"i18next-resources-to-backend": "1.2.1",
		"next": "15.1.0",
		"next-auth": "4.24.11",
		"nuqs": "2.2.3",
		"react": "19.0.0",
		"react-dom": "19.0.0",
		"react-i18next": "15.2.0",
		"server-only": "0.0.1",
		"sonner": "1.7.1",
		"usehooks-ts": "3.1.0",
		"zod": "3.24.1"
	},
	"devDependencies": {
		"@eslint/eslintrc": "3.2.0",
		"@next/bundle-analyzer": "15.1.0",
		"@tanstack/eslint-plugin-query": "5.62.1",
		"@tanstack/react-query-devtools": "5.62.7",
		"@types/node": "22.10.2",
		"@types/react": "19.0.1",
		"@types/react-dom": "19.0.2",
		"@typescript-eslint/eslint-plugin": "8.18.1",
		"autoprefixer": "10.4.20",
		"cross-env": "7.0.3",
		"eslint": "9.17.0",
		"eslint-config-next": "15.1.0",
		"eslint-config-prettier": "9.1.0",
		"eslint-plugin-react-compiler": "19.0.0-beta-a7bf2bd-20241110",
		"postcss": "8.4.49",
		"prettier": "3.4.2",
		"prettier-plugin-tailwindcss": "0.6.9",
		"tailwind-variants": "0.3.0",
		"tailwindcss": "3.4.16",
		"typescript": "5.7.2"
	},
	"pnpm": {
		"overrides": {
			"@types/react": "19.0.1",
			"@types/react-dom": "19.0.2"
		}
	}
}
```

Next-config:

```
import type { NextConfig } from 'next';

const nextConfig: NextConfig = {
	basePath: '/next',
	output: 'standalone',
	experimental: {
		// Testing with different packages to optimize to see if it improves.
		optimizePackageImports: [
			'@nivo/core',
			'@nivo/pie',
			'@nivo/bar',
			'@nextui-org',
			'@nextui-org/react',
			'@nextui-org/theme',
			'@nextui-org/system',
			'@nextui-org/calendar',
			'@nextui-org/autocomplete',
			'@nextui-org/alert',
			'@nextui-org/avatar',
			'@nextui-org/badge',
			'@nextui-org/breadcrumbs',
			'@nextui-org/button',
			'@nextui-org/card',
			'@nextui-org/checkbox',
			'@nextui-org/checkbox',
			'@nextui-org/chip',
			'@nextui-org/progress',
			'@nextui-org/code',
			'@nextui-org/date-input',
			'@nextui-org/date-picker',
			'@nextui-org/date-picker',
			'@nextui-org/divider',
			'@nextui-org/dropdown',
			'@nextui-org/drawer',
			'@nextui-org/accordion',
			'@nextui-org/form',
			'@nextui-org/image',
			'@nextui-org/input',
			'@nextui-org/input-otp',
			'@nextui-org/kbd',
			'@nextui-org/link',
			'@nextui-org/listbox',
			'@nextui-org/modal',
			'@nextui-org/navbar',
			'@nextui-org/pagination',
			'@nextui-org/popover',
			'@nextui-org/progress',
			'@nextui-org/radio',
			'@nextui-org/calendar',
			'@nextui-org/scroll-shadow',
			'@nextui-org/select',
			'@nextui-org/skeleton',
			'@nextui-org/slider',
			'@nextui-org/snippet',
			'@nextui-org/spacer',
			'@nextui-org/spinner',
			'@nextui-org/switch',
			'@nextui-org/table',
			'@nextui-org/tabs',
			'@nextui-org/input',
			'@nextui-org/date-input',
			'@nextui-org/tooltip',
			'@nextui-org/user',
			'@wt/components',
			'@internationalized/date',
			'i18next-browser-languagedetector',
			'framer-motion',
			'sonner',
			'usehooks-ts',
			'react-i18next',
			'i18next',
			'nuqs',
			'@react-spring/web',
			'@react-aria',
			'@react-aria/calendar',
			'loadash',
			'react-aria-components',
			'@adobe/react-spectrum',
		],
	},
	logging: {
		fetches: {
			fullUrl: true,
		},
	},
};

export default nextConfig;
```

**Reply by icyJoseph** [reply]
Thanks! Yeh I was able to reproduce this with a minimal example. I'll try to investigate further today.

**Reply by J4v4Scr1pt** [reply]
If it matters we are using pnpm throughout the product.

**Reply by icyJoseph** [reply]
That's not needed, I've identified the PR that, most likely, introduced this side-effect, https://github.com/vercel/next.js/pull/71202

- Next.js 15.0.0-canary.187 -> No issue
- Next.js 15.0.0-canary.188 -> Issue begins

There are only 3 commits on that release cycle, https://github.com/vercel/next.js/releases/tag/v15.0.0-canary.188, so it has to #71202

**Reply by J4v4Scr1pt** [reply]
I can confirm that by using 15.0.0-canary.187 the values goes back to normal 🎉.
What a great find!

Do you have any suggestion with this?
Ignore and use latest and wait for fix or use canary.187 until fix?

**Reply by icyJoseph** [reply]
Now the question is, is this the right behaviour?

I have a Next 13 app, where the FIrst Load JS is not zero...

**Reply by icyJoseph** [reply]
But it looks like the app router Route Handlers, had no first load JS, so maybe this is actually a fix, the pages router API Routes did have a non zero value, that's what I see.

**Reply by icyJoseph** [reply]
Maybe you need to tag a core maintainer and they can give us some insight? @samcx perhaps? or @unstubbable :) ? 

Please read through the last 5 comments or so, in short, https://github.com/vercel/next.js/pull/71202, with this PR, First Load JS (shown in the build output) for Route Handlers, went from 0 in v14, to some number in v15. That's the expected behavior right?

**Reply by J4v4Scr1pt** [reply]
Sure I have bad knowledge on who I should tag 😅. Don't want to bother anyone unmercenary 😊.
But Sam that  you tagged could be enough for starters it seems :).

I wonder why my pages also go up by ~20kbs using latest NextJs version 🤔 
Because they go back down to almost Next14 and React18 numbers using the 15.0.0-canary.187. 
I was counting on some increase because of React 19 so that is fine, on 15.0.0-canary.187 the pages are up with ~5-10kb.

**Reply by icyJoseph** [reply]
React has gotten larger, IIRC: https://www.reddit.com/r/reactjs/comments/1h7gpwe/comment/m0m3d9s/?share_id=rjYva1pcb3Pp6ibv9sB-H maybe that's it? 

```console
React itself is very small, react-dom is the heavy one. If you do the vite starter template

18.3.1 : 143.90 kB │ gzip: 46.34 kB

19.0.0 186.49 kB │ gzip: 58.96 kB

```

Let's try to avoid spamming now, and wait for a response from the tags

**Reply by samcx** [reply]
@icyJoseph Thanks for investigating and taking a look!

Pinging the team on this—

**Reply by samcx** [reply, marked as answer]
To enable `use cache` in Route Handlers, we started generating client reference manifests and client chunks for Route Handlers in [#70897](https://github.com/vercel/next.js/pull/70897). This led to the shared initial client JS size being also shown for Route Handlers. It's confusing for users because from the output it's not immediately clear that this is client JS, which is irrelevant for Route Handlers.

We should just print a dash there! The 0 B in v14 was similarly strange, I think.

**Reply by J4v4Scr1pt** [reply]
Ahh ok, thanks for this information @samcx  🙏!  
I think I understand, both 0b and in my case 109kb are just numbers not meaningful to the route handler itself.
I other words both are equally correct and can safely be ignored, correct? :)

**Reply by samcx** [reply]
Yep!

**Reply by J4v4Scr1pt** [reply]
Thank you both @icyJoseph and @samcx for helping me understand this 🙌. Marked your answer as the definitive answer samcx.
Have a superb day!

**Comment by MobinAskari** [comment]
Wow this was really confusing to me as well!
